core/sdk-core/src/main/java/software/amazon/awssdk/core/waiters/Waiter.java

```diff
@@ -0,0 +1,134 @@
+     * @return the response
+     */
+    WaiterResponse<T> run(Supplier<T> pollingFunction);
+
+    /**
+     * Creates a newly initialized builder for the waiter object.
+     *
+     * @param responseClass the response class
+     * @param <T> the type of the response
+     * @return a Waiter builder
+     */
+    static <T> Builder<T> builder(Class<? extends T> responseClass) {
+        return DefaultWaiter.builder();
+    }
+
+    /**
+     * The Waiter Builder
+     * @param <T> the type of the resource
+     */
+    interface Builder<T> {
+        /**
+         * Defines a list of {@link WaiterAcceptor}s to check whether an expected state has met after executing an operation.
+         *
+         * <p>
+         * The SDK will iterate over the acceptors list and the first acceptor to match the result of the operation transitions
+         * the waiter to the state specified in the acceptor.
+         *
+         * <p>
+         * This completely overrides any WaiterAcceptor currently configured in the builder via
+         * {@link #addAcceptor(WaiterAcceptor)}
+         *
+         * @param waiterAcceptors the waiter acceptors
+         * @return a reference to this object so that method calls can be chained together.
+         */
+        Builder<T> acceptors(List<WaiterAcceptor<T>> waiterAcceptors);
+
+        /**
+         * Adds a {@link WaiterAcceptor} to the end of the ordered waiterAcceptors list.
+         *
+         * <p>
+         * The SDK will iterate over the acceptors list and the first acceptor to match the result of the operation transitions
+         * the waiter to the state specified in the acceptor.
+         *
+         * @param waiterAcceptors the waiter acceptors
+         * @return a reference to this object so that method calls can be chained together.
+         */
+        Builder<T> addAcceptor(WaiterAcceptor<T> waiterAcceptors);
+
+        /**
+         * Defines a {@link PollingStrategy} to use when polling a resource
+         *
+         * @param pollingStrategy the polling strategy to use
+         * @return a reference to this object so that method calls can be chained together.
+         */
+        Builder<T> pollingStrategy(PollingStrategy pollingStrategy);
+
+        /**
+         * Defines a {@link PollingStrategy} to use when polling a resource
+         *
+         * @param pollingStrategy the polling strategy to use
+         * @return a reference to this object so that method calls can be chained together.
+         */
+        default Builder<T> pollingStrategy(Consumer<PollingStrategy.Builder> pollingStrategy) {
+            PollingStrategy.Builder builder = PollingStrategy.builder();
+            pollingStrategy.accept(builder);
+            return pollingStrategy(builder.build());
+        }
+
+        /**
+         * Defines the {@link ScheduledExecutorService} used to schedule async polling attempts
+         * Only required if you are calling {@link Waiter#runAsync(Supplier)}
```

Comment:
Again, "only required during creation, if you invoke this method during usage" feels like a gross invariant. Wouldn't it be better to encode that into the type? (`AsyncWaiter` versus `Waiter`)